metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,200 @@
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N](
+    ledgerTree: Ref[F, LedgerTree],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    OptionT(updateLedger(block)).flatMapF { ledger =>
+      val hashValid = ledger.hash == block.header.postStateHash
+      validateTransactions(block.body, ledger).map {
+        _.map(_ && hashValid)
+      }
+    }.value
+  }
+
+  private def validateTransactions(
+      body: Block.Body,
+      ledger: Ledger
+  ): F[Option[Boolean]] = {
+    //TODO: Validate transactions PM-3131/3132
+    true.some.pure[F]
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    updateLedger(block).flatMap {
+      case Some(ledger) =>
+        if (block.hash != commitQC.blockHash)
+          saveState(block.header, ledger).as(true)
+        else
+          for {
+            //TODO: PM-3110:
+            // chkpOpt <- constructCheckpoint(ledger, commitQC)
+            _ <- saveState(block.header, ledger)
+            //TODO: PM-3110:
+            // _ <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+          } yield true
+
+      case None =>
+        Sync[F].raiseError(
+          //TODO: better error info with Either?
+          new IllegalStateException(s"Could not execute block: ${block.hash}")
+        )
+    }
+  }
+
+  private def updateLedger(block: Block): F[Option[Ledger]] = {
+    (for {
+      ledgers    <- ledgerTree.get
+      execHeight <- lastExecutedHeader.get.map(_.height)
+    } yield {
+      def loop(block: Block): OptionT[F, Ledger] = {
+        def doUpdate(ledger: Ledger) =
+          OptionT.liftF(updateLedgerByBlock(ledger, block))
+
+        ledgers.get(block.header.parentHash) match {
+          case Some((oldLedger, _)) =>
+            doUpdate(oldLedger)
+
+          case None if block.header.height <= execHeight =>
+            OptionT.none
+
+          case None =>
+            for {
+              //TODO: no reason to look up storage for execution; validation???
```

Comment:
Yeah, it was just meant for myself to start the discussion 😄 

The way that `BlockExecutor` works it shouldn't be necessary to ever look up block storage when running `executeBlock`. What about validation though?

Blocks that are behind the last committed block will be rejected based on height anyway. But perhaps it may happen that we will be validating a descendant of last committed block whose parent has not been validated/executed before. In such case it would make sense to keep the storage look-up, but I'm curious about your input.

Reply:
So you are saying that when the class is created, you do 1 lookup in persistent ledger storage, put it in the in-memory ledger tree, then executions just build on that one, and only do writes to the persistent ledger store. You are wondering if we would ever need to read the persistent state store for validation. 

>  But perhaps it may happen that we will be validating a descendant of last committed block whose parent has not been validated/executed before.

It is indeed possible that we miss a (few) round(s) for any reason, and we need to validate a block that is the descendant of the block we know about as last committed, but we have some missing in the chain. In this case the `SyncService` would download these but not call validation on them, as their validity is proven by the High Q.C. extended by the proposed block we were given. The missing ancestor blocks, up to the High Q.C. would be persisted, and `validateBlock` called on the proposed child. 

Your algorithm should walk up the tree until it finds the state of the last committed block in memory. I don't see where it could find anything in persistent storage along the way that it doesn't have in memory. 

Are we talking about the same thing?

Reply:
> Are we talking about the same thing?

Not quite 😉, but I wasn't clear about which storage I was talking about. I meant specifically the `blockStorage` lookup that occurs in `projectLedger`. It is not needed for execution because `BlockExecutor` will take care of providing blocks, but based on your comment above, I believe we are in an agreement that it is still needed to facilitate fully asynchronous validation. Comment removed